### PR TITLE
fix issue 19933 and 18816: MSVC: Undefined std{in,out,err} with -betterC

### PR DESCRIPTION
**Pull request by rainers**
moved C runtime check into template function

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]
Otherwise looks good.

File: src/core/stdc/stdio.d [hunk 1]
```diff
@@ -897,12 +904,45 @@ else version (CRuntime_Microsoft)
...
+    {
+        immutable void* _nullfunc = null;
+        template isAvailable(alias f)
+        {
+            version (X86)
+                enum cPrefix = "_";
+            else
+                enum cPrefix = "";
+
+            mixin(`pragma(linkerDirective, "/alternatename:` ~ cPrefix ~ f.mangleof ~ `=` ~ cPrefix ~ _nullfunc.mangleof ~ `");`);
```
[the remark this answers is not in the document]
Comment (rainers) [reply src/core/stdc/stdio.d:921]: Hmm, I just copied it from msvc.c. Not sure if linkerDirective supports CTFE evaluation, I'll try.

[the remark this answers is not in the document]
Comment (thewilsonator) [reply src/core/stdc/stdio.d:921]: Fair enough. FWIW, I'd consider it a bug if you can only pass string literals and not CTFE'd strings, if that turns out to be the case, please do file one.

[the remark this answers is not in the document]
Comment (rainers) [reply src/core/stdc/stdio.d:921]: Indeed, mixin is unnecessary, I removed it.

**Comment by rainers** [comment]
> Well, VS2013 has been EOL for coming up 3 years now. I don't see any reason to go out of our way to support it.

It still has extended support for another 2 years, but I think this fix doesn't have to target VS2013. It should't break existing functionality (without -betterC) though, so I'll move part of the code back to msvc.d...

**Comment by rainers** [comment]
> so I'll move part of the code back to msvc.d...

Much simpler now.

File: src/core/stdc/stdio.d (hunk 1 above)
[the remark this answers is not in the document]
Comment (thewilsonator) [reply src/core/stdc/stdio.d]: Thanks for looking into that

**Comment by rainers** [comment]
Tests pass now. They uncovered that the default mingw libraries correspond to VS 2003, so they don't work here out of the box. Maybe it's time to change the default to the newest available (msvcrt120 corresponds to VS 2015).

**Comment by thewilsonator** [comment]
> They uncovered that the default mingw libraries correspond to VS 2003

Please tell me thats a that a typo and was meant to be 2013?

**Comment by rainers** [comment]
> Please tell me thats a that a typo and was meant to be 2013?

It wasn't a typo because I thought `msvcrt` was the default, but indeed it is `msvcrt120` which corresponds to VS 2013. This doesn't include the required symbol `__acrt_iob_func`, though. 

What's needed is `vcruntime140.lib` that corresponds to VS 2015. It depends on `ucrtbase.lib` which the definitions for universal CRT (which might be a hassle to install on Windows 7 or earlier).

It's not so so easy to specify these libraries and switch to lld-link and the mingw libraries on a system that has VS installed (as they don't start with `msvcrt`). I've updated the test in a way that should work in both environments.

**Review by RazvanN7 — APPROVED** [review]

**Comment by WalterBright** [comment]
@rainers Thank you!

**Comment by ibuclaw** [comment]
This PR introduced a testsuite regression on Solaris
https://issues.dlang.org/show_bug.cgi?id=22829

**Pull request merged**